### ISSUE TITLE
FR: Adjustable color contrast on e-ink

### ISSUE THREAD

**Issue by brainrom** [koreader/koreader#10336]
**Does your feature request involve difficulty completing a task? Please describe.**
1. Add UI (probably radio buttons with few values) to change contrast multiplier
2. Find the code block, where image rendering happens
3. Apply contrast multiplier to every image pixel

**Describe the solution you'd like**
Colors on e-ink display should become more contrast.

**Describe alternatives you've considered**
Tried to do a CSS stylesheet with `filter: contrast()`, filter are not applied.

**Additional context**
I'm using PocketBook 613 Color. Stock reader (PB Reader) does a color correction, which leads to more natural and more contrast colors. Koreader displays images as is, and colors looks faded. First screenshot from KOReader, second - from PB Reader. I'm familiar with Lua, successfully configured development env, but can't find exact place where I should add color correction. If someone points me, I think I can implement this feature yourself, at least with configuration via Advanced settings.
![scr0004](https://user-images.githubusercontent.com/15963835/232340237-1175d4bb-94e8-422f-bc53-2e8f348c87de.jpg)
![scr0003](https://user-images.githubusercontent.com/15963835/232340240-a157d0c1-2a26-409c-90f2-1158523883ce.jpg)

**Comment by Frenzie** [comment]
See <https://github.com/koreader/koreader/pull/9756>, pinging @rjd22.

What does it say under help → report a bug in detected device?

**Comment by rjd22** [comment]
@brainrom honestly the first picture looks better to me. The colors seems way more balanced. Maybe a bit bland so a bit more salutation would be better.

Could you adjust the value of following line on your device to 2.0 and make a screenshot?

https://github.com/koreader/koreader/blob/99265ba1b7061ae0521be8e777db8bdfbf59c2dc/frontend/device/pocketbook/device.lua#L567

**Comment by offset-torque** [comment]
> honestly the first picture looks better to me. The colors seems way more balanced. Maybe a bit bland so a bit more salutation would be better.

I thought the same **@rjd22** and out of curiosity looked up the device. This photo is from a review and left side is this color reader:
![pocketbook color](https://user-images.githubusercontent.com/84420582/232499209-392472eb-2c41-4bdd-9666-f32bebdab2fe.jpg)



All the colors looks washed out as we can see. Probably stock reader increases saturation to compensate for this fact to make colors look more vivid. 

Maybe taking a screenshot is not the correct approach here. I suspect that over saturated picture looks normal-ish in real life. If the colors are really pale like this, maybe an adjustable saturation setting is a good idea for these type of screens.

**Comment by rjd22** [comment]
@offset-torque I need examples of different images with different saturation settings to decide on a correct setting for the Pocketbook Color as I don't have the device myself.

This involves using the stock pocketbook images and some color comics with the following saturation settings:

- 1.5
- 2.0
- 2.5
- 3.0

This way we can get a good baseline on what saturation is the most correct for this device.

**Comment by brainrom** [comment]
I tried increase `color_saturation` constant, but it looks like it doesn't affect at all. Also tried to set `color_saturation` and `canHWDither` for `PocketBook` class, no luck.

**Comment by rjd22** [comment]
> I tried increase `color_saturation` constant, but it looks like it doesn't affect at all. Also tried to set `color_saturation` and `canHWDither` for `PocketBook` class, no luck.

Did you reboot koreader or the pocketbook after doing so? What is the model number of your pocketbook that is communicated to KoReader?

**Comment by brainrom** [comment]
Enabled dithering and picture becomes saturated....
Maybe this should be enabled by default?

**Comment by rjd22** [comment]
> Enabled dithering and picture becomes saturated.... Maybe this should be enabled by default?

Good to hear it's now working for you. Could you do some testing with the different saturation settings and different pictures to see what setting is appropriate for the PBColor? 

Please make photos of your screen on the brightest setting so we can see what the results are and compare them.

**Comment by brainrom** [comment]
What reference picture I should use?

**Comment by rjd22** [comment]
> What reference picture I should use?

The one with the blue water and boats, and the one with the houses with a green hill on the right are good ones from the stock photos. Also do some comic pages as those often need a different setting. Feel free to give your personal opinion next to the images ;).

**Comment by NiLuJe** [comment]
> Enabled dithering

That's the expected behavior on PDFs/CBZs, it's a user-toggle, and it's opt-in (because CBZs are a niche use-case in the grand scheme of things, so off makes more sense as a default).

**Comment by rjd22** [comment]
> > Enabled dithering
> 
> FWIW, that's the expected behavior on PDFs/CBZs, it's a user-toggle, and it's opt-in (because CBZs are a niche use-case in the grand scheme of things, so off makes more sense as a default).
> 
> ePubs _will_ detect image content automatically, though.

Especially if you take into account that most manga is black/white and doesn't really need dithering to be on. Having dithering on decreases the performance of the images rendering. I only enable it when really needed.

**Comment by grabi678** [comment]
I use InkPad Color and I can report that `color_saturation = 1.5` with dithering enabled works fine with most PDF and JPG files. 
But this is not applied to EPUB files! Colors still remain pale and faded and I could not find an option to enable dithering for EPUB documents or something similar.

**Comment by NiLuJe** [comment]
> ePubs _will_ detect image content automatically, though.

**Comment by Frenzie** [comment]
Font color and background color and such through CSS wouldn't be affected by that, would they?

**Comment by NiLuJe** [comment]
Nope, just the actual images that take a sufficient amount of screen real-estate.

**Comment by grabi678** [comment]
> But this is not applied to EPUB files! Colors still remain pale and faded and I could not find an option to enable dithering for EPUB documents or something similar.

Ok. I tried to proof my assertion with screenshots and photos of PDF and EPUB files with the same picture but I can't. There is no evident difference. Maybe the images in the EPUB I have read where pale themselves.

**Comment by ckilb** [comment]
I've got the Pocketbook Inkpad Color 2.
In PDFs you can enable dithering in the bottom menu.

For some reason, this isn't implemented in Epubs.

For a quick solution you can edit ffi/framebuffer_pocketbook.lua and find all conditions where dithering is checked (if dither then ...) and remove it (or simply set dither = true before).

I did so and it works flawlessly on my device.

Now I wonder how to proceed here. Sure, we could make the dithering option available for Epubs, too. But from my point of view the option does not make sense, at least not on Pocketbook devices. Also, it's quite hard to find. I wasn't even aware of it (and nearly starting to accepting bad colors on KOReader) until I found out about it.
Why should one be able to decide if they want to crappy image colors instead of satured ones?

I tend to implement a new flag isDitheringEnabled in device.lua which can be

nil => the logic keeps acting like it is now
yes => dithering is always enabled, the option to disable it is removed
no => dithering is always disabled, the option to enable it is removed
For InkpadColor and InkpadColor 2 we could set the option to yes.
Logic for other devices (Kobo) will be unchanged.

**Comment by NiLuJe** [comment]
> For some reason, this isn't implemented in Epubs.

It is, it happens automatically if there's a sufficiently large image being displayed.

(The intent being you really don't want to dither *everything*, because 99.999% of your screen refreshes will be plain text).

**Comment by NiLuJe** [comment]
I can't quite see anything obviously wrong anywhere, so I suspect it's likely a snafu somewhere along the line because of the 24bpp fb used on those, so I'm curious as to what actually happens on PDFs..

Because we *don't* support dithering on 24bpp fbs. Which is why i'm curious as to what actually happens in PDFs, because the "Dithering" button in there is a magical beast that will do either SW or HW ;). And there isn't technically any HW dithering on PB, we piggy-backed the saturation thing on top of it, but that's not really dithering.

**Comment by NiLuJe** [comment]
TL;DR: Moar logz.

**Comment by NiLuJe** [comment]
Nope, can't see anything "color" specific in the CRe codepaths that could explain this, so, yes, please send verbose debug logs of reading picture books (;p) so I can confirm whether or not we get a `dither` flag set.

And perhaps stick a few debug prints after https://github.com/koreader/koreader/blob/d350418367ddf39d752d05e0587e562d7d4af2c4/frontend/document/credocument.lua#L785-L786 to log the returned values...

**Comment by rjd22** [comment]
> I can't quite see anything obviously wrong anywhere, so I suspect it's likely a snafu somewhere along the line because of the 24bpp fb used on those, so I'm curious as to what actually happens on PDFs..
> 
> Because we _don't_ support dithering on 24bpp fbs. Which is why I'm curious as to what actually happens in PDFs, because the "Dithering" button in there is a magical beast that will do either SW or HW depending on the device capabilities (and a dev setting if the device can do both) ;). And there isn't technically any HW dithering on PB, we piggy-backed the color saturation thing on top of it, but that's not really dithering.

The problem is that the dithering boolean is not passed when dithering is off so we cannot use them to check if the content is a image or plain text.

Color saturation is a heavy task and slows down performance considerably so running it on every refresh would be a no go. Honestly I would really want to tackle the following points to have a good control on this:

- be able to detect if the rendered content contains an color image that needs saturation
- be able to control the amount of saturation and vibrance in the client to allow users to adjust the setting for their own needs

**Comment by ckilb** [comment]
> Nope, can't see anything "color" specific in the CRe codepaths that could explain this, so, yes, please send verbose debug logs of reading picture books (;p) so I can confirm whether or not we get a `dither` flag set.
> 
> And perhaps stick a few debug prints after
> 
> https://github.com/koreader/koreader/blob/d350418367ddf39d752d05e0587e562d7d4af2c4/frontend/document/credocument.lua#L785-L786
> 
> to log the returned values...

So, after removed my dither = true statements again it's still working.
This is weird because I'm sure it wasn't working in the past.
For once a cover image (nearly full screen had bad colors) but also an inline image that consists just of black and red colors. Now both seem to be working without any issue.

@rjd22 you say color saturation is a heavy task, but is this the case for Pocketbooks? The original software seems to dither everything without any battery issues.
Also, the Pocketbook dithering is using some inkview.h method, so if I understand correctly the logic that is actually used has been implemented by Pocketbook themselves. Even with color saturation always enabled the CPU seems to be less stressed for each page turn in the original Software than in KOReader. If you turn on the LED color indication setting in Pocketbook you can see that the LED flashes in a shorter time using the original software compared to KOReader. But yeah, basically this is just speculation as we don't know how the LED indication is exactly implemented.

**Comment by NiLuJe** [comment]
> The problem is that the dithering boolean is not passed when dithering is off so we cannot use them to check if the content is a image or plain text.

Err, which `dithering` option being off are we talking about, and on what kind of documents?

**Comment by rjd22** [comment]
> > The problem is that the dithering boolean is not passed when dithering is off so we cannot use them to check if the content is a image or plain text.
> 
> Err, which `dithering` option being off are we talking about, and on what kind of documents?
> 
> (It feels like you're talking CBZ/PDFs, while we were discussing ePubs ;)).

Ah yes, I meant CBZ's sorry for the confusion.

**Comment by dmalinovsky** [comment]
I believe this is fixed with https://github.com/koreader/koreader-base/pull/1751.
